### PR TITLE
feat: add support for Flat Config

### PR DESCRIPTION
**Pull request by susisu**
- About Flat Config: https://eslint.org/docs/latest/use/configure/configuration-files-new
- See https://github.com/hatena/eslint-config-hatena/blob/flat-config/docs/usage.md for the detailed usage

### CODE REVIEW

File: src/config/index.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import tsEslint from 'typescript-eslint';
+import rules from '../rules';
+
+export type ConfigOptions = Readonly<
+  Partial<{
+    /**
+     * TypeScript の設定ファイル (languageOptions.parserOptions.project)
+     * デフォルト: ./tsconfig.json
+     */
+    tsProject: boolean | string | string[] | undefined;
```
Comment (mizdra) [inline src/config/index.ts:17]: Off topic: I think it is a good idea to make the default value of `tsProject` to `true`.

https://typescript-eslint.io/blog/parser-options-project-true/

File: src/config/index.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    {
+      files: ['**/*.{js,jsx,mjs}'],
+      languageOptions: {
+        // 現代では type="script" な環境で JS を書くことはまずないので、
+        // デフォルトで type="module" なJSであるとして lint する
+        sourceType: 'module',
+      },
+    },
+    {
+      files: ['**/*.cjs'],
```
Comment (mizdra) [inline src/config/index.ts:83]: MAY: Maybe you should add `cts`.

```suggestion
      files: ['**/*.{cjs,cts}'],
```

File: src/config/index.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      languageOptions: {
+        sourceType: 'commonjs',
+      },
+    },
+    {
+      files: ['**/*.{ts,tsx,cts,mts}'],
+      languageOptions: {
+        sourceType: 'module',
+        parser: tsEslint.parser as Linter.FlatConfigParserModule,
+        parserOptions: tsProject ? { project: tsProject } : {},
```
Comment (mizdra) [inline src/config/index.ts:93]: MAY: It seems that `tsProject  === false` or `tsProject === ''` will result in `{}`. Is that intended?

Wouldn't the following code be better?

```suggestion
        parserOptions: tsProject ?? { project: tsProject },,
```

**Review by mizdra — APPROVED** [review]
I add some comments. However, I think it is ready to be merged 👍.

File: src/config/index.ts (hunk 1 above)
Comment (susisu) [reply src/config/index.ts:17]: Yep, I know.

The reason I used `./tsconfig.json` here is that the current `.eslintrc` style config uses `./tsconfig.json` and I wanted to keep the configs as same as possible.

https://github.com/hatena/eslint-config-hatena/blob/547609d55009e084cca0b78ca7d54f87523724ae/src/classic/typescript.ts#L21

File: src/config/index.ts (hunk 2 above)
Comment (susisu) [reply src/config/index.ts:83]: `.cts` files are compiled into CommonJS (`.cjs`) files, but they themselves are modules.

File: src/config/index.ts (hunk 3 above)
Comment (susisu) [reply src/config/index.ts:93]: It's intended. You can unset `parserOptions.project` by setting a falsy value if you don't use type-aware rules.
(The config enables type-aware rules by default, so it's very uncommon though.)

Maybe we can disable type-aware rules when `tsProject` is false in the future version (as I do in [my own config](https://github.com/susisu/eslint-config/blob/08cbfbc0dff8151b56fa24aaaeddbd8e27cfd340/src/config/index.ts#L96)).

**Comment by susisu** [comment]
Thanks!

**Pull request merged**